### ISSUE TITLE
[no-restricted-syntax]: allow naming each rule in the array

### ISSUE THREAD

**Issue by johan** [eslint/eslint#14081]
**What rule do you want to change?** `no-restricted-syntax`

**Does this change cause the rule to produce more or fewer warnings?** Neither.

**How will the change be implemented? (New option, new default behavior, etc.)?** An optional `name` property beside `selector` and `message`.

**Please provide some example code that this change will affect:**

Scenario:
* your organization has a shared set of eslint rules that each project imports via `extends`
* several of those rules are implemented by way of `no-restricted-syntax`
* some of those rules are there as a critical early warning system flagging preventable security breaches from dangerous apis like `window.open` that can cause code execution side effects if used unwisely
* others suggest mere code style best practices like using a centralized logger instead of `console.log` variants
* some projects pulling in the centrally maintained ruleset want to disable one of these rules (say, the `use-alternate-logger` one), while retaining all the others

Sadly, they can't, as these rules don't have names you can address and change severity level for, your only option is to bite down and disable the rule locally in the code everywhere you have a `console.log`, or disable the whole set, including the security breach warnings, via

```js
"no-restricted-syntax": 0
```

If we could name each of these rules in the config, we could instead retain all the security rules and disable mere specific ones that don't apply to us:

```json
"no-restricted-syntax/use-alternate-logger": 0
```

**What does the rule currently do for this code?** N/A

**What will the rule do after it's changed?** Prevent disasters where eslint coverage provides on-going audits against accidentally tripping unsafe apis (or whatever else your `no-restricted-syntax` rules do)

**Are you willing to submit a pull request to implement this change?** No. I presume this request might also require some re-architecturing of the eslint plugin system, to accommodate configuration creating new entries in the rules namespace.

**Comment by mdjermanovic** [comment]
Hi @johan, thanks for the issue! 

> ```json
> "no-restricted-syntax/use-alternate-logger": 0
> ```

This looks like a significant change in ESLint's core design that might be useful for just one rule.

Have you considered creating a plugin with separate rules? Rules that can be enforced with `no-restricted-syntax` would be very simple.

For example, if you have:

```js
{
    "rules": {
        "no-restricted-syntax": [
            "error",
            {
                "selector": "BinaryExpression[operator='in']",
                "message": "Don't use 'in' operator."
            }
        ]
    }
}
```

an equivalent rule would have code like this:

```js
module.exports = {
    meta: {
        type: "problem",
        schema: []
    },
    create(context) {
        return {
            "BinaryExpression[operator='in']"(node) {
                context.report({ node, message: "Don't use 'in' operator." })
            }
        }
    }
};
```

**Comment by nzakas** [comment]
I also made a package that makes simple rules like these easy to make:
https://github.com/humanwhocodes/eslint-simple-rule

**Comment by nzakas** [comment]
We won’t be moving forward with this, so closing.
